### ISSUE TITLE
"parallel input mismatch" when "from *" added to front of forked/join query

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#4085]
Repro is from a community/zync user of Zui Insiders, so based on timing it is presumed that the repro is with Zed commit 9c0f097 (or something close to that).

The user shared a sophisticated query that's being withheld from this public issue due to its proprietary nature. However, of potential significance is the fact that it included `fork` and `join` pieces. The user indicated this query worked as expected in isolation. However, they then attempted to issue the same query across all pools in the Zed lake lake by using the `from *` in front of the query, which resulted in the following error message:

![image](https://user-images.githubusercontent.com/5934157/189985619-5d970361-2e8e-4bbe-ada2-786be7ee484d.png)

**Comment by philrz** [comment]
FYI while the fix is still being worked on, I was hacking around trying to repro with a simpler query and couldn't produce the same error message, but as @nwt explained, what I observed shares a root cause with what the community user experienced.

My attempt was to base a repro on the examples in the [Zed `join` tutorial](https://zed.brimdata.io/docs/tutorials/join).

```
$ zed -version
Version: v1.2.0-49-g9c0f0973

$ curl http://localhost:9867/version
{"version":"v1.2.0-49-g9c0f0973"}

$ zed ls
people.ndjson 2Ej1jPGGEPEJAiwQg7eS8tcyWoz key ts order desc
fruit.ndjson 2Ej1k95ZqocRjE8Qa8RNl2tn92o key ts order desc

$ zed query 'from people.ndjson'
{name:"morgan",age:61,likes:"tart"}
{name:"jessie",age:30,likes:"plain"}
{name:"quinn",age:14,likes:"sweet",note:"many kids enjoy sweets"}
{name:"chris",age:47,likes:"tart"}

$ zed query 'from fruit.ndjson'
{name:"strawberry",color:"red",flavor:"sweet"}
{name:"avocado",color:"green",flavor:"savory"}
{name:"banana",color:"yellow",flavor:"sweet"}
{name:"dates",color:"brown",flavor:"sweet",note:"in season"}
{name:"apple",color:"red",flavor:"tart"}
{name:"figs",color:"brown",flavor:"plain"}

$ cat forked.zed 
from * | fork (
  => has(color) | sort flavor
  => has(age) | sort likes
) | inner join on flavor=likes eater:=name
```

What I observed is that it only works sometimes.

```
$ while true; do zed query -I forked.zed; echo "just ran the query" ; done
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
{name:"figs",color:"brown",flavor:"plain",eater:"jessie"}
{name:"strawberry",color:"red",flavor:"sweet",eater:"quinn"}
{name:"banana",color:"yellow",flavor:"sweet",eater:"quinn"}
{name:"dates",color:"brown",flavor:"sweet",note:"in season",eater:"quinn"}
{name:"apple",color:"red",flavor:"tart",eater:"morgan"}
{name:"apple",color:"red",flavor:"tart",eater:"chris"}
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
just ran the query
{name:"figs",color:"brown",flavor:"plain",eater:"jessie"}
{name:"strawberry",color:"red",flavor:"sweet",eater:"quinn"}
{name:"banana",color:"yellow",flavor:"sweet",eater:"quinn"}
{name:"dates",color:"brown",flavor:"sweet",note:"in season",eater:"quinn"}
{name:"apple",color:"red",flavor:"tart",eater:"morgan"}
{name:"apple",color:"red",flavor:"tart",eater:"chris"}
just ran the query
just ran the query
...
```

I won't try to transcribe the explanation for how the `from` / `fork` combo's current behavior is causing this. Suffice to say, an effective workaround is to put a `sort` between the two.

```
$ cat forked-plus-sort.zed
from * | sort | fork (
  => has(color) | sort flavor
  => has(age) | sort likes
) | inner join on flavor=likes eater:=name

$ while true; do zed query -I forked-plus-sort.zed; echo "just ran the query" ; done
{name:"figs",color:"brown",flavor:"plain",eater:"jessie"}
{name:"banana",color:"yellow",flavor:"sweet",eater:"quinn"}
{name:"dates",color:"brown",flavor:"sweet",note:"in season",eater:"quinn"}
{name:"strawberry",color:"red",flavor:"sweet",eater:"quinn"}
{name:"apple",color:"red",flavor:"tart",eater:"chris"}
{name:"apple",color:"red",flavor:"tart",eater:"morgan"}
just ran the query
{name:"figs",color:"brown",flavor:"plain",eater:"jessie"}
{name:"banana",color:"yellow",flavor:"sweet",eater:"quinn"}
{name:"dates",color:"brown",flavor:"sweet",note:"in season",eater:"quinn"}
{name:"strawberry",color:"red",flavor:"sweet",eater:"quinn"}
{name:"apple",color:"red",flavor:"tart",eater:"chris"}
{name:"apple",color:"red",flavor:"tart",eater:"morgan"}
just ran the query
{name:"figs",color:"brown",flavor:"plain",eater:"jessie"}
{name:"banana",color:"yellow",flavor:"sweet",eater:"quinn"}
{name:"dates",color:"brown",flavor:"sweet",note:"in season",eater:"quinn"}
{name:"strawberry",color:"red",flavor:"sweet",eater:"quinn"}
{name:"apple",color:"red",flavor:"tart",eater:"chris"}
{name:"apple",color:"red",flavor:"tart",eater:"morgan"}
just ran the query
...
```

**Comment by philrz** [comment]
Verified in Zed commit 3d32d97.

Circling back to repro, I've since recognized why I wasn't able to reproduce the same error message that the community user reported. The missing ingredient was to add yet another pool beyond the ones I wanted to join. So revisiting the repro with 9c0f097, now I've got a third pool which contains the [sample.zng](https://github.com/brimdata/brim/blob/main/src/test/shared/data/sample.zng) data from the Brim repo.

```
$ zed -version
Version: v1.2.0-49-g9c0f0973

$ curl http://localhost:9867/version
{"version":"v1.2.0-49-g9c0f0973"}

$ zed ls
people.ndjson 2F0LUwg94Nt0w35X7pEO2JHnT3n key ts order desc
fruit.ndjson 2F0LVYBdPfP2La9LceJF5eMTKgg key ts order desc
sample.zng 2F0LVwivM0HIs3FIZEQc6TJTjcd key ts order desc

$ zed query -I forked.zed
status code 500: parallel input mismatch: 3 parents with 2 flowgraph paths
```

Now at Zed commit 3d32d97 that has the fixes from #4087:

```
$ zed -version
Version: v1.2.0-59-g3d32d974

$ curl http://localhost:9867/version
{"version":"v1.2.0-59-g3d32d974"}

$ zed ls
sample.zng 2F0LqAnYiFjhRFcrvY8KbQiMuqM key ts order desc
people.ndjson 2F0LpEq8tcQsXuDFvTL3U1syhxg key ts order desc
fruit.ndjson 2F0Lpj7MIpKOOaOmOT8YZ1FzXWl key ts order desc

$ zed query -I forked.zed
{name:"figs",color:"brown",flavor:"plain",eater:"jessie"}
{name:"strawberry",color:"red",flavor:"sweet",eater:"quinn"}
{name:"banana",color:"yellow",flavor:"sweet",eater:"quinn"}
{name:"dates",color:"brown",flavor:"sweet",note:"in season",eater:"quinn"}
{name:"apple",color:"red",flavor:"tart",eater:"morgan"}
{name:"apple",color:"red",flavor:"tart",eater:"chris"}
```

The fixed Zed is also present in the latest Zui Insiders build 74, so the original community user should be able to move past this issue now.

![image](https://user-images.githubusercontent.com/5934157/191126344-37c821dc-dbfd-44cf-b98b-2ec374f193db.png)

Thanks @nwt!
